### ISSUE TITLE
TrailingCommaInArguments incorrectly wants a trailing comma in single arg hash

### ISSUE THREAD

**Issue by ptarjan** [rubocop-hq/rubocop#2846]
`TrailingCommaInArguments` with this config:

```
Style/TrailingCommaInArguments:
  EnforcedStyleForMultiline: comma
```

wants this code

```
foo({
  bar: 1,
})
```

to be this

```
foo({
  bar: 1,
},)
```

which seems pretty wrong.

**Comment by lamont-granquist** [comment]
if you're gonna go with the multiline 'comma' style you should probably write it like:

```ruby
foo(
  {
    bar: 1,
  },
)
```

or:

```ruby
foo(
  { bar: 1 },
)
```

its a multiline argument so i think its just doing you what you told it to.

**Comment by ptarjan** [comment]
@lamont-granquist Yes, I agree. Your two formats this Cop behaves well, but my formatting also seems reasonable too. Why does the trailing comma cop care about how I do my hashes? Another example might help
```
# good
foo({
  bar: 1,
  baz: 2,
})

# what the cop does
foo({
  bar: 1,
  baz: 2,
},)
```

**Comment by lamont-granquist** [comment]
It isn't your hashes, its the argument.  your parens in foo() are on different lines so they're considered 'multiline' so it inserts the trailing comma because that's exactly what you told it to do.

**Comment by ptarjan** [comment]
@lamont-granquist How can I tell it not to put that weird comma there? I basically never want `},)` to be recommended.

**Comment by lamont-granquist** [comment]
drop `EnforcedStyleForMultiline: comma` and just accept the default style.  that option works better for `Style/TrailingCommaInLiterals` and not so good for `Style/TrailingCommaInArguments`

**Comment by ptarjan** [comment]
But the default for `EnforcedStyleForMultiline` is `no_comma` which isn't what my codebase wants. Do you intend to make the `comma` option work better for it?

**Comment by lamont-granquist** [comment]
I'm not a rubocop dev, I'm just a user working through roughly the same problem.

https://github.com/chef/chef/commit/92bec4bb96d857f027068fc56af46c76c484fd50

That is what it looks like when you use `no_comma` and it eliminates the `},)` weirdness that nobody wants.  Not sure what you mean by isn't what your 'codebase' wants...

**Comment by alexdowad** [comment]
@ptarjan Can you make an argument that the way the `comma` style works is *different from what is stated by the documentation in default.yml*? Just saying it "seems pretty wrong" doesn't help.

The fact is that we have had lots of complaints about these cops in the past... and everybody seems to have a different opinion about how they *should* work.

If the current behavior is contrary to the documentation, then it is buggy and will be fixed. If not, then we could look at adding another optional style. *But*, that new style should have a concise definition in prose, describing exactly when comma should or should not be present. (Just code examples is not enough.)

**Comment by lamont-granquist** [comment]
Then only ideas I could think of would be adding a `EnforcedStyleForMultiline:  comma_unless_singular` style -- or maybe doing something to enforce that in multiline arguments, hash arguments should not have closing brackets on the opening or closing line of the argument (basically try to auto-fix arguments to look like my examples up there).

I've got a huge pile of 'meh' for either alternative, though, I just went with :comma for literals and :no_comma for arguments and stopped worrying about it...  It looks like that was why the TrailingComma cop was broken out into the two cops so they could be tweaked independently.

I don't see any bugs in how it behaves with respect to the docs...

**Comment by bbatsov** [comment]
There hasn't been much activity on this ticket and our Core Team is spread too thin on so many tasks, so I'll just close it for the sake of having a cleaner lists of tasks to focus on. We'd gladly review a PR, but it's unlikely that'd we're going to tackle this ourselves in the foreseeable future.
